From: Stephen Gildea <gildea@stop.mail-abuse.org>
To: Matthias Klose <doko@cs.tu-berlin.de>
Cc: ian@caliban.org
Subject: Re: Bug#262105: /etc/bash_completion: bad fn name, global var set 
Date: Sun, 01 Aug 2004 17:43:40 -0400

>   please could you update the patch against the new bash_completion
>   version and CC it to Ian?

Sure, here's a diff against release 20040711:

--- bash_completion.20040711	2004-07-11 14:36:39
+++ bash_completion	2004-08-01 17:28:33
@@ -160,9 +160,7 @@ 
 #
 have()
 {
-	unset -v have
-	PATH=$PATH:/sbin:/usr/sbin:/usr/local/sbin type $1 &>/dev/null &&
-		have="yes"
+	PATH=$PATH:/sbin:/usr/sbin:/usr/local/sbin type $1 &>/dev/null
 }
 
 # use GNU sed if we have it, since its extensions are still used in our code
@@ -532,8 +530,8 @@ { have service || [ -d /etc/init.d/ ]; }
 	fi
 
 	return 0
-}
-[ -n "${have:-}" ] && complete -F _service service
+} &&
+complete -F _service service
 [ -d /etc/init.d/ ] && complete -F _service $default \
 	$(for i in /etc/init.d/*; do echo ${i##*/}; done)
 
@@ -1074,8 +1072,8 @@ [ $UNAME = Linux ] && have ipsec &&
 	esac
 
 	return 0
-}
-[ $UNAME = Linux ] && [ -n "${have:-}" ] && complete -F _ipsec ipsec
+} &&
+complete -F _ipsec ipsec
 
 # Postfix completion.
 #
@@ -1821,7 +1819,7 @@ have apt-get &&
 		remove)
 			if [ -f /etc/debian_version ]; then
 				# Debian system
-				COMPREPLY=( $( _comp-dpkg-installed-packages \
+				COMPREPLY=( $( _comp_dpkg_installed_packages \
 						$cur ) )
 			else
 				# assume RPM based
@@ -1876,8 +1874,8 @@ have apt-get &&
 
 
 	return 0
-}
-[ -n "${have:-}" ] && complete -F _apt_get $filenames apt-get
+} &&
+complete -F _apt_get $filenames apt-get
 
 # Debian apt-cache(8) completion.
 #
@@ -1947,8 +1945,8 @@ have apt-cache &&
 
 
 	return 0
-}
-[ -n "${have:-}" ] && complete -F _apt_cache $filenames apt-cache
+} &&
+complete -F _apt_cache $filenames apt-cache
 
 
 # Debian aptitude(1) completion
@@ -1993,11 +1991,11 @@ have aptitude && {
 		   return 0
 		   ;;
 	       @(purge|remove|reinstall|forbid-version))
-  		   COMPREPLY=( $( _comp-dpkg-installed-packages $cur ) )
+  		   COMPREPLY=( $( _comp_dpkg_installed_packages $cur ) )
 		   return 0
 		   ;;
 	       unhold)
-  		   COMPREPLY=( $( _comp-dpkg-hold-packages $cur ) )
+  		   COMPREPLY=( $( _comp_dpkg_hold_packages $cur ) )
 		   return 0
 		   ;;
 
@@ -2037,7 +2035,7 @@ have aptitude && {
 
 	return 0
 }
-[ -n "${have:-}" ] && complete -F _aptitude $default aptitude
+complete -F _aptitude $default aptitude
 }
 
 # Debian apt-build(1) completion.
@@ -2064,7 +2062,7 @@ have apt-build &&
 			return 0
 			;;
 		remove)
-			COMPREPLY=( $( _comp-dpkg-installed-packages \
+			COMPREPLY=( $( _comp_dpkg_installed_packages \
 					$cur ) )
 			return 0
 			;;
@@ -2103,8 +2101,8 @@ have apt-build &&
 
 
 	return 0
-}
-[ -n "${have:-}" ] && complete -F _apt_build $filenames apt-build
+} &&
+complete -F _apt_build $filenames apt-build
 
 # chsh(1) completion
 #
@@ -2161,8 +2159,8 @@ have chkconfig &&
 			_services
 		fi
 	fi
-}
-[ -n "${have:-}" ] && complete -F _chkconfig chkconfig
+} &&
+complete -F _chkconfig chkconfig
 
 # This function provides simple user@host completion
 #
@@ -2443,8 +2441,8 @@ have rsync &&
 	esac
  
 	return 0
-}
-[ -n "${have:-}" ] && complete -F _rsync $nospace rsync
+} &&
+complete -F _rsync $nospace rsync
 
 # Linux route(8) completion
 #
@@ -2563,8 +2561,8 @@ have make &&
 				{split($1,A,/ /);for(i in A)print A[i]}' \
 				$makef 2>/dev/null | command grep "^$cur" ))
 	fi
-}
-[ -n "${have:-}" ] && complete -F _make $filenames make gmake pmake
+} &&
+complete -F _make $filenames make gmake pmake
 
 # GNU tar(1) completion
 #
@@ -2652,8 +2650,8 @@ have jar &&
 			_filedir
 			;;
 	esac
-}
-[ -n "${have:-}" ] && complete -F _jar $filenames jar
+} &&
+complete -F _jar $filenames jar
 
 # Linux iptables(8) completion
 #
@@ -2713,8 +2711,8 @@ have iptables &&
 		;;
 	esac
 
-} 
-[ -n "${have:-}" ] && complete -F _iptables iptables
+} &&
+complete -F _iptables iptables
 
 # tcpdump(8) completion
 #
@@ -2745,8 +2743,8 @@ have tcpdump &&
 			-E' -- $cur ) )
 	fi
 
-}
-[ -n "${have:-}" ] && complete -F _tcpdump tcpdump
+} &&
+complete -F _tcpdump tcpdump
 
 # autorpm(8) completion
 #
@@ -2762,8 +2760,8 @@ have autorpm &&
 				   auto add fullinfo info help install list \
 				   remove set' -- $cur ) )
 
-}
-[ -n "${have:-}" ] && complete -F _autorpm autorpm
+} &&
+complete -F _autorpm autorpm
 
 # This meta-cd function observes the CDPATH variable, so that cd additionally
 # completes on directories under those specified in CDPATH.
@@ -2941,8 +2939,8 @@ have ant &&
 			    $( awk -F'"' '/<target [^n]/ {if ($1 ~ /name=/) { print $2 } else if ($3 ~ /name=/) {print $4} else if ($5 ~ /name=/) {print $6}}' \
 				$buildfile | grep "^$cur" ) )
 	fi
-}
-[ -n "${have:-}" ] && complete -F _ant $filenames ant
+} &&
+complete -F _ant $filenames ant
 
 have nslookup &&
 _nslookup()
@@ -2956,8 +2954,8 @@ have nslookup &&
 			       srchlist= defname search port= querytype= \
 			       type= recurse retry root timeout vc \
 			       ignoretc' -- $cur ) )
-}
-[ -n "${have:-}" ] && complete -F _nslookup nslookup
+} &&
+complete -F _nslookup nslookup
 
 # mysqladmin(1) completion
 #
@@ -2989,8 +2987,8 @@ have mysqladmin &&
 				   password ping processlist reload refresh \
 				   shutdown status variables version' \
 		       -- $cur ) )
-}
-[ -n "${have:-}" ] && complete -F _mysqladmin mysqladmin
+} &&
+complete -F _mysqladmin mysqladmin
 
 # gzip(1) completion
 #
@@ -3031,8 +3029,8 @@ have gzip &&
 
 	COMPREPLY=( $( compgen -f -X "$xspec" -- $cur ) \
 		    $( compgen -d -- $cur ) )
-}
-[ -n "${have:-}" ] && complete -F _gzip $filenames gzip
+} &&
+complete -F _gzip $filenames gzip
 
 # bzip2(1) completion
 #
@@ -3069,8 +3067,8 @@ have bzip2 &&
 
 	COMPREPLY=( $( compgen -f -X "$xspec" -- $cur ) \
 		    $( compgen -d -- $cur ) )
-}
-[ -n "${have:-}" ] && complete -F _bzip2 $filenames bzip2
+} &&
+complete -F _bzip2 $filenames bzip2
 
 # openssl(1) completion
 #
@@ -3098,8 +3096,8 @@ have openssl &&
 	fi
 
 	return 0
-}
-[ -n "${have:-}" ] && complete -F _openssl $default openssl
+} &&
+complete -F _openssl $default openssl
 
 # screen(1) completion
 #
@@ -3139,8 +3137,8 @@ have screen &&
 	fi
 
 	return 0
-}
-[ -n "${have:-}" ] && complete -F _screen $default screen
+} &&
+complete -F _screen $default screen
 
 # lftp(1) bookmark completion
 #
@@ -3158,8 +3156,8 @@ have lftp &&
 	fi
 
 	return 0
-}
-[ -n "${have:-}" ] && complete -F _lftp $default lftp
+} &&
+complete -F _lftp $default lftp
 
 # ncftp(1) bookmark completion
 #
@@ -3177,8 +3175,8 @@ have ncftp &&
 	fi
 
 	return 0
-}
-[ -n "${have:-}" ] && complete -F _ncftp $default ncftp
+} &&
+complete -F _ncftp $default ncftp
 
 # gdb(1) completion
 #
@@ -3200,8 +3198,8 @@ have gdb &&
 				awk '{if ($1 ~ /^'$prev'/) print $2}' ) )" \
 				-- "$cur" ) )
 	fi
-}
-[ -n "${have:-}" ] && complete -F _gdb $filenames gdb
+} &&
+complete -F _gdb $filenames gdb
 
 # Postgresql completion
 #
@@ -3413,8 +3411,8 @@ have gcc &&
 	else
 		_filedir
 	fi
-}
-[ -n "${have:-}" ] && complete $filenames -F _gcc gcc g++ c++ g77 gcj gpc
+} &&
+complete $filenames -F _gcc gcc g++ c++ g77 gcj gpc
 [ $UNAME = GNU -o $UNAME = Linux -o $UNAME = Cygwin ] && \
 complete $filenames -F _gcc cc
 
@@ -3433,13 +3431,13 @@ have cardctl &&
 					   resume reset eject insert scheme' \
 			       -- $cur ) )
 	fi
-}
-[ -n "${have:-}" ] && complete -F _cardctl cardctl
+} &&
+complete -F _cardctl cardctl
 
 # This function is required by _dpkg() and _dpkg-reconfigure()
 #
 have dpkg && {
-_comp-dpkg-installed-packages()
+_comp_dpkg_installed_packages()
 {
 	grep -A 2 "Package: $1" /var/lib/dpkg/status | \
 		grep -B 2 'ok installed' | grep "Package: $1" | cut -d\  -f2
@@ -3485,7 +3483,7 @@ have dpkg && {
 		return 0
 		;;
 	-@(r|L|P|-@(remove|purge|listfiles)))
-		COMPREPLY=( $( _comp-dpkg-installed-packages $cur ) )
+		COMPREPLY=( $( _comp_dpkg_installed_packages $cur ) )
 		return 0
 		;;
 	*)
@@ -3554,10 +3552,10 @@ have dpkg-reconfigure &&
 				       -u --unseen-only -h --help -s --showold \
 				       --force --terse' -- $cur ) )
 	else
-	    COMPREPLY=( $( _comp-dpkg-installed-packages $cur ) )
+	    COMPREPLY=( $( _comp_dpkg_installed_packages $cur ) )
 	fi
-}
-[ -n "${have:-}" ] && complete -F _dpkg_reconfigure $default dpkg-reconfigure
+} &&
+complete -F _dpkg_reconfigure $default dpkg-reconfigure
 
 # Debian Linux dselect(8) completion.
 #
@@ -3592,8 +3590,8 @@ have dselect &&
 
 
 	return 0
-}
-[ -n "${have:-}" ] && complete -F _dselect $filenames dselect
+} &&
+complete -F _dselect $filenames dselect
 
 # Java completion
 #
@@ -3808,8 +3806,8 @@ have javadoc &&
 		# packages completion
 		_java_packages
 	fi
-}
-[ -n "${have:-}" ] && complete -F _javadoc $filenames javadoc
+} &&
+complete -F _javadoc $filenames javadoc
 
 # javac completion
 #
@@ -3843,8 +3841,8 @@ have javac &&
 		# source files completion
 		_filedir java
 	fi
-}
-[ -n "${have:-}" ] && complete -F _javac $filenames javac
+} &&
+complete -F _javac $filenames javac
 
 # PINE address-book completion
 #
@@ -3858,8 +3856,8 @@ have pine &&
 
 	COMPREPLY=( $( compgen -W '$( awk "{print \$1}" ~/.addressbook 2>/dev/null)' \
 			-- $cur ) )
-}
-[ -n "${have:-}" ] && complete -F _pineaddr $default pine
+} &&
+complete -F _pineaddr $default pine
 
 
 # mutt completion
@@ -4085,8 +4083,8 @@ have reportbug &&
 	    		$( apt-cache pkgnames -- $cur ) )
 	_filedir
 	return 0
-}
-[ -n "${have:-}" ] && complete -F _reportbug $filenames reportbug
+} &&
+complete -F _reportbug $filenames reportbug
 
 # Debian querybts(1) completion
 #
@@ -4118,8 +4116,8 @@ have querybts &&
 			-s --source -w --web -u --ui --interface \
 			wnpp boot-floppies' -- $cur ) \
 	    		$( apt-cache pkgnames -- $cur ) )
-}
-[ -n "${have:-}" ] && complete -F _querybts $filenames querybts
+} &&
+complete -F _querybts $filenames querybts
 
 # update-alternatives completion
 #
@@ -4258,8 +4256,8 @@ have python &&
 
 
 	return 0
-}
-[ -n "${have:-}" ] && complete -F _python $filenames python
+} &&
+complete -F _python $filenames python
 
 # Perl completion
 #
@@ -4409,8 +4407,8 @@ have rcs &&
 	# default to files if nothing returned and we're checking in.
 	# otherwise, default to directories
 	[ ${#COMPREPLY[@]} -eq 0 -a $1 = ci ] && _filedir || _filedir -d
-}
-[ -n "${have:-}" ] && complete -F _rcs $filenames ci co rlog rcs rcsdiff
+} &&
+complete -F _rcs $filenames ci co rlog rcs rcsdiff
 
 # lilo(8) completion
 #
@@ -4458,8 +4456,8 @@ have lilo &&
 			-M -p -P -q -r -R -s -S -t -T -u -U -v -V -w -x -z' -- \
 			$cur ) )
 	fi
-}
-[ -n "${have:-}" ] && complete -F _lilo lilo
+} &&
+complete -F _lilo lilo
 
 # links completion
 #
@@ -4495,8 +4493,8 @@ have links &&
 	esac
   
 	return 0
-}
-[ -n "${have:-}" ] && complete -F _links $filenames links
+} &&
+complete -F _links $filenames links
 
 [ $UNAME = FreeBSD ] && {
 # FreeBSD package management tool completion
@@ -4564,8 +4562,8 @@ have portupgrade &&
 	COMPREPLY=( ${COMPREPLY[@]%-*} )
 
 	return 0
-}
-[ -n "${have:-}" ] && complete -F _portupgrade $dirnames portupgrade
+} &&
+complete -F _portupgrade $dirnames portupgrade
 
 # FreeBSD portinstall completion
 #
@@ -4593,8 +4591,8 @@ have portinstall &&
 	COMPREPLY=( ${COMPREPLY[@]} ${COMPREPLY2[@]} )
 
 	return 0
-}
-[ -n "${have:-}" ] && complete -F _portinstall $dirnames portinstall
+} &&
+complete -F _portinstall $dirnames portinstall
 
 # Slackware Linux removepkg completion
 #
@@ -4607,8 +4605,8 @@ have removepkg && [ -f /etc/slackware-ve
 	cur=${COMP_WORDS[COMP_CWORD]}
 
 	COMPREPLY=( $( (cd /var/log/packages; compgen -f -- "$cur") ) )
-}
-[ -n "${have:-}" ] && complete -F _removepkg $filenames removepkg &&
+} &&
+complete -F _removepkg $filenames removepkg &&
 	complete $dirnames -f -X '!*.tgz' installpkg upgradepkg explodepkg
 
 # look(1) completion
@@ -4624,8 +4622,8 @@ have look && 
 	if [ $COMP_CWORD = 1 ]; then
 		COMPREPLY=( $( compgen -W '$(look $cur)' ) )
 	fi
-}
-[ -n "${have:-}" ] && complete -F _look $default look
+} &&
+complete -F _look $default look
 
 # ypcat(1) and ypmatch(1) completion
 #
@@ -4652,8 +4650,8 @@ have ypmatch &&
 	fi
 
 	return 0
-}
-[ -n "${have:-}" ] && complete -F _ypmatch ypmatch ypcat
+} &&
+complete -F _ypmatch ypmatch ypcat
 
 # mplayer(1) completion
 #
@@ -4984,8 +4982,8 @@ 			# if you don't have installed mplayer
 
 	return 0
 }
-}
-[ -n "${have:-}" ] && complete $filenames -F _mplayer mplayer mencoder gmplayer
+} &&
+complete $filenames -F _mplayer mplayer mencoder gmplayer
 
 # KDE dcop completion
 #
@@ -5002,8 +5000,8 @@ have dcop &&
 	    compstr=$( command echo ${COMP_WORDS[*]} | sed "s/ $cur$//" )
 	fi
 	COMPREPLY=( $( compgen -W '$( command $compstr | sed s/\(.*\)// )'  -- $cur ) )
-}
-[ "${have:-}" ] && complete -F _dcop dcop
+} &&
+complete -F _dcop dcop
 
 # wvdial(1) completion
 #
@@ -5049,8 +5047,8 @@ have wvdial &&
 			;;
 	esac
 
-}
-[ -n "${have:-}" ] && complete -F _wvdial wvdial
+} &&
+complete -F _wvdial wvdial
 
 # gpg(1) completion
 #
@@ -5087,8 +5085,8 @@ have gpg &&
 				-q -n -N $(gpg --dump-options)' -- $cur ) )
 	 fi
 
-}
-[ "${have:-}" ] && complete -F _gpg $default gpg
+} &&
+complete -F _gpg $default gpg
 
 # iconv(1) completion
 #
@@ -5114,8 +5112,8 @@ have iconv &&
 	fi
 		
 	return 0
-}
-[ -n "${have:-}" ] && complete -F _iconv $default iconv
+} &&
+complete -F _iconv $default iconv
 
 # dict(1) completion
 #
@@ -5334,8 +5332,8 @@ have mkisofs &&
 		_filedir
 	fi
 
-}
-[ -n "${have:-}" ] && complete -F _mkisofs mkisofs
+} &&
+complete -F _mkisofs mkisofs
 
 # mc(1) completion
 #
@@ -5353,8 +5351,8 @@ have mc &&
 	else
 		_filedir -d
 	fi
-}
-[ "${have:-}" ] && complete -F _mc $filenames mc
+} &&
+complete -F _mc $filenames mc
 
 # yum(8) completion
 #
@@ -5416,8 +5414,8 @@ have yum && {
 						grouplist ' -- $cur ) )
 		;;
 	esac
-}
-[ -n "${have:-}" ] && complete -F _yum $filenames yum
+} &&
+complete -F _yum $filenames yum
 
 # yum-arch(8) completion
 #
@@ -5844,8 +5842,8 @@ have dd &&
          COMPREPLY=( $( compgen -W '--help --version' -- $cur ) \
                      $( compgen -W 'bs cbs conv count ibs if obs of seek 
 skip' -S '=' -- $cur ) )
-}
-[ "${have:-}" ] && complete -F _dd $nospace dd
+} &&
+complete -F _dd $nospace dd
 
 # CUPS cancel(1) completion
 #
@@ -5858,8 +5856,8 @@ have cancel &&
 	cur=${COMP_WORDS[COMP_CWORD]}
 
 	COMPREPLY=( $( lpstat | cut -d' ' -f1 | grep "^$cur" ) )
-}
-[ "${have:-}" ] && complete -F _cancel $filenames cancel
+} &&
+complete -F _cancel $filenames cancel
 
 # aspell(1) completion
 #
@@ -5981,8 +5979,8 @@ have xmms &&
 
 	fi
 
-}
-[ "${have:-}" ] && complete -F _xmms $filenames xmms
+} &&
+complete -F _xmms $filenames xmms
 
 # info(1) completion
 #
